Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -253,3 +253,181 @@ theorem Language.IsContextFree.reverse (L : Language T) :
+  corresponding_rules : ∀ r : ContextFreeRule T g₀.NT, r ∈ g₀.rules → r.lift liftNT ∈ g.rules
+  /-- Each rule of the bigger grammar whose input nonterminal the smaller grammar recognizes
+      has a corresponding rule in the smaller grammar. -/
+  preimage_of_rules :
+    ∀ r : ContextFreeRule T g.NT,
+      (r ∈ g.rules ∧ ∃ n₀ : g₀.NT, liftNT n₀ = r.input) →
+        (∃ r₀ ∈ g₀.rules, r₀.lift liftNT = r)
+
+lemma LiftedContextFreeGrammar.sinkNT_inverse_liftNT (G : LiftedContextFreeGrammar T) :
+    ∀ x : G.g.NT, (∃ n₀, G.sinkNT x = some n₀) → (Option.map G.liftNT (G.sinkNT x) = x) := by
+  intro x ⟨n₀, hx⟩
+  rw [hx, Option.map_some']
+  apply congr_arg
+  by_contra hnx
+  cases (G.sinkNT_liftNT n₀ ▸ G.sink_inj x (G.liftNT n₀)) hx with
+  | inl case_valu => exact hnx case_valu.symm
+  | inr case_none => exact Option.noConfusion (hx ▸ case_none)
+
+lemma LiftedContextFreeGrammar.lift_produces {G : LiftedContextFreeGrammar T}
+    {w₁ w₂ : List (Symbol T G.g₀.NT)} (hG : G.g₀.Produces w₁ w₂) :
+    G.g.Produces (Symbol.liftString G.liftNT w₁) (Symbol.liftString G.liftNT w₂) := by
+  rcases hG with ⟨r, rin, hr⟩
+  rcases hr.exists_parts with ⟨u, v, bef, aft⟩
+  refine ⟨r.lift G.liftNT, G.corresponding_rules r rin, ?_⟩
+  rw [ContextFreeRule.rewrites_iff]
+  use Symbol.liftString G.liftNT u, Symbol.liftString G.liftNT v
+  constructor
+  · simpa only [Symbol.liftString, List.map_append] using congr_arg (Symbol.liftString G.liftNT) bef
+  · simpa only [Symbol.liftString, List.map_append] using congr_arg (Symbol.liftString G.liftNT) aft
+
+/-- Derivation by `G.g₀` can be mirrored by `G.g` derivation. -/
+lemma LiftedContextFreeGrammar.lift_derives {G : LiftedContextFreeGrammar T}
+    {w₁ w₂ : List (Symbol T G.g₀.NT)} (hG : G.g₀.Derives w₁ w₂) :
+    G.g.Derives (Symbol.liftString G.liftNT w₁) (Symbol.liftString G.liftNT w₂) := by
+  induction hG with
+  | refl => rfl
+  | tail _ orig ih => exact ih.trans_produces (lift_produces orig)
+
+/-- A `Symbol` is good iff it is one of those nonterminals that result from sinking or it is any
+terminal. -/
+def LiftedContextFreeGrammar.GoodLetter {G : LiftedContextFreeGrammar T} : Symbol T G.g.NT → Prop
+  | Symbol.terminal _ => True
+  | Symbol.nonterminal n => ∃ n₀ : G.g₀.NT, G.sinkNT n = n₀
+
+/-- A string is good iff every `Symbol` in it is good. -/
+def LiftedContextFreeGrammar.GoodString {G : LiftedContextFreeGrammar T}
+    (s : List (Symbol T G.g.NT)) : Prop :=
+  ∀ a ∈ s, GoodLetter a
+
+lemma LiftedContextFreeGrammar.sink_produces {G : LiftedContextFreeGrammar T}
+    {w₁ w₂ : List (Symbol T G.g.NT)} (hG : G.g.Produces w₁ w₂) (hw₁ : GoodString w₁) :
+    G.g₀.Produces (Symbol.sinkString G.sinkNT w₁) (Symbol.sinkString G.sinkNT w₂) ∧
+      GoodString w₂ := by
+  rcases hG with ⟨r, rin, hr⟩
+  rcases hr.exists_parts with ⟨u, v, bef, aft⟩
+  rcases G.preimage_of_rules r (by
+      constructor
+      · exact rin
+      rw [bef] at hw₁
```

Comment:
I suggest moving ` rw [bef] at hw₁` to line 363 (before the last `rcases`), and removing it here and at line 401, because `hw₁` is never used without rewriting `bef` at it.